plugins/modem-manager/fu-mm-device.c

```diff
@@ -931,6 +1116,170 @@ fu_mm_device_write_firmware_mbim_qdu (FuDevice *device, GBytes *fw, GError **err
+					    (FuDeviceLockerFunc) fu_mm_device_firehose_open,
+					    (FuDeviceLockerFunc) fu_mm_device_firehose_close,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+
+	progress_signal_id = g_signal_connect_swapped (self->firehose_updater, "write-percentage",
+						       G_CALLBACK (fu_device_set_progress), self);
+
+	write_result = fu_firehose_updater_write (self->firehose_updater,
+						  rawprogram_silo,
+						  rawprogram_actions,
+						  error);
+
+	g_signal_handler_disconnect (self->firehose_updater, progress_signal_id);
+
+	return write_result;
+}
+
+static gboolean
+fu_mm_setup_firmware_dir (FuMmDevice *self, GError **error)
+{
+	g_autofree gchar *cachedir = NULL;
+	g_autofree gchar *mm_fw_dir = NULL;
+
+	/* create a directory to store firmware files for modem-manager plugin */
+	cachedir = fu_common_get_path (FU_PATH_KIND_CACHEDIR_PKG);
+	mm_fw_dir = g_build_filename (cachedir, "modem-manager", "firmware", NULL);
+	if (g_mkdir_with_parents (mm_fw_dir, 0700) == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "Failed to create '%s': %s",
+			     mm_fw_dir, g_strerror (errno));
+		return FALSE;
+	}
+
+	if (!fu_common_set_firmware_search_path (mm_fw_dir, error))
+		return FALSE;
+
+	self->firmware_path = g_steal_pointer (&mm_fw_dir);
+
+	return TRUE;
+}
+
+static gboolean
+fu_mm_copy_firehose_prog (FuMmDevice *self, FuArchive *archive, GError **error)
+{
+	g_autofree gchar *qcom_fw_dir = NULL;
+	g_autofree gchar *firehose_prog_path = NULL;
+	g_autoptr(GBytes) firehose_prog = NULL;
+
+	/* lookup firehose-prog bootloader */
+	firehose_prog = fu_archive_lookup_by_fn (archive, "firehose-prog.mbn", error);
+	if (firehose_prog == NULL)
+		return FALSE;
+
+	qcom_fw_dir = g_build_filename (self->firmware_path, "qcom", NULL);
+	if (!fu_common_mkdir_parent (qcom_fw_dir, error))
+		return FALSE;
+
+	firehose_prog_path = g_build_filename (qcom_fw_dir, "prog_firehose_sdx24.mbn", NULL);
+
+	if (!fu_common_set_contents_bytes (firehose_prog_path, firehose_prog, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_mm_device_write_firmware_firehose (FuDevice *device, GBytes *fw, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE (device);
+
+	gboolean res;
+	GBytes *firehose_rawprogram;
+	g_autofree gchar *old_fw_path = NULL;
+	g_autoptr(FuArchive) archive = NULL;
+	g_autoptr(XbSilo) firehose_rawprogram_silo = NULL;
+	g_autoptr(GPtrArray) firehose_rawprogram_actions = NULL;
+
+	/* decompress entire archive ahead of time */
+	archive = fu_archive_new (fw, FU_ARCHIVE_FLAG_IGNORE_PATH, error);
+	if (archive == NULL)
+		return FALSE;
+
+	old_fw_path = fu_common_get_firmware_search_path (NULL);
+	/* firehose modems that switch to the EDL mode over QCDM port require
+	 * firehose binary to be present in the firmware-loader search path. */
+	if (!fu_mm_setup_firmware_dir (self, error))
+		return FALSE;
+	if (!fu_mm_copy_firehose_prog(self, archive, error))
+		return FALSE;
```

Comment:
I think you need to restore the old path in case of failure here.

Reply:
you are right, fixed it